src/rhsm/connection.py

```diff
@@ -670,19 +673,38 @@ def close_connection(self) -> None:
         Try to close connection to server
         :return: None
         """
-        if self.__conn is not None:
-            # Do proper TLS shutdown handshake (TLS tear down) first
-            if self.__conn.sock is not None:
-                log.debug(f"Closing HTTPS connection {self.__conn.sock}")
-                try:
-                    self.__conn.sock.unwrap()
-                except ssl.SSLError as err:
-                    log.debug(f"Unable to close TLS connection properly: {err}")
-                else:
-                    log.debug("TLS connection closed")
-            # Then it is possible to close TCP connection
+        if self.__conn is None:
+            return
+
+        if self.__conn.sock is None:
+            log.debug(f"Closing TCP connection to {self.__conn.host}")
             self.__conn.close()
-        self.__conn = None
+            self.__conn = None
+            return
+
+        # The server may not send the `close_notify` alert when it closes the
+        # connection on its side.
+        # Here we set a timeout equal to three times the measured response time
+        # to prevent the connection from waiting until TCP timeout.
+        # See RHEL-17345.
+        log.debug(f"Closing TLS connection {self.__conn.sock}")
+
+        response_time: float = self.smoothed_rt or 0.5
+        timeout_time: int = math.ceil(response_time * 3)
+
+        def on_timeout(signum, frame) -> None:
+            raise TimeoutError(f"Did not get response in {timeout_time}s")
+
+        signal.signal(signalnum=signal.SIGALRM, handler=on_timeout)
+        signal.alarm(timeout_time)
+
+        try:
+            self.__conn.sock.unwrap()
+        except (ssl.SSLError, TimeoutError) as err:
+            log.debug(f"TLS connection could not be closed: {err}")
```

Comment:
The alarm should be canceled in the `else` statement using `signal.alarm(0)`. See: https://man7.org/linux/man-pages/man2/alarm.2.html

Also, not sure how will it work with multi-threaded `rhsm.service`.

Reply:
I would also add original debug message to else statement.

Reply:
After a bit of investigating I found out that connection reuse is disabled for the D-Bus server and this code will never run in multi-threaded context.
https://github.com/candlepin/subscription-manager/blob/47d2729b537a39ccc6d9c61407ea4e1720b02b8f/src/rhsmlib/dbus/server.py#L84-L88

Reply:
I think that this code could be run in multi-threaded context.

When reusing of connection is disabled, then it only means that each REST API request uses it's own TLS connection, but the connection still have to be closed.

Correct me if I'm wrong, but I believe that there could be more threads doing REST API calls, when there are two apps using rhsm.service concurrently.

Reply:
The whole setup does not seem thread-safe. The `__conn` is class attribute, it would be shared between the threads. If we want to make it thread-safe we'd need to change the architecture of the whole connection reuse.

I'm still not sure this PR makes sense. We're not handling "the sad path" on so many other places this seems unnecessary -- we know the initial cause will be fixed server-side and this patch will not be executed in real conditions.